examples/wmt/input_pipeline.py

```diff
@@ -0,0 +1,255 @@
+                     dynamic_batching=True,
+                     max_target_length=512,
+                     max_eval_target_length=512):
+  """Load and return dataset of batched examples for use during training."""
+  if batch_size % n_devices:
+    raise ValueError("Batch size %d isn't divided evenly by n_devices %d" %
+                     (batch_size, n_devices))
+  if vocab_path is None:
+    vocab_path = os.path.expanduser('~/vocab.subwords')
+
+  train_data = tfds.load(
+      name=dataset_name,
+      split=tfds.Split.TRAIN,
+      data_dir=data_dir,
+      shuffle_files=True)
+
+  eval_data = tfds.load(
+      name=dataset_name,
+      split=tfds.Split.VALIDATION,
+      data_dir=data_dir,
+      shuffle_files=False)
+  features_info = tfds.builder(dataset_name, data_dir=data_dir).info
+  keys = ([features_info.supervised_keys[0]],
+          [features_info.supervised_keys[1]])
+
+  # TFDS's supervised keys looks like: (('en',), ('de',),)
+  # we want to map data to 'inputs' and 'targets' features.
+  # e.g. {'en': example_in, 'de': example_out} to
+  #      {'inputs': example_in, 'targets': example_out}
+  to_features_dict = (
+      lambda x: {'inputs': x[keys[0][0]], 'targets': x[keys[1][0]]})
+  train_data = train_data.map(to_features_dict)
+  eval_data = eval_data.map(to_features_dict)
+
+  try:
+    subword_tokenizer = load_tfds_subword_vocab(vocab_path)
+  except tf.errors.NotFoundError:
+    logging.info('Subword vocab not found, building shared subword dictionary.')
+    def corpus_generator():
+      for data in train_data:
+        yield data['inputs'].numpy()
+        yield data['targets'].numpy()
+    subword_tokenizer = tfds.features.text.SubwordTextEncoder.build_from_corpus(
+        corpus_generator(),
+        target_vocab_size=target_vocab_size,
+        max_corpus_chars=max_corpus_chars)
+    # save vocab
+    save_tfds_subword_vocab(subword_tokenizer, vocab_path)
+
+  # Encode strings with subword tokenizer.
+  def encode(lang1, lang2):
+    res1 = subword_tokenizer.encode(
+        lang1.numpy()) + [subword_tokenizer.vocab_size]  # Add EOS Token
+    res2 = subword_tokenizer.encode(
+        lang2.numpy()) + [subword_tokenizer.vocab_size]  # Add EOS Token
+    return res1, res2
+
+  def tf_encode(x):
+    src, tgt = x['inputs'], x['targets']
+    result_src, result_tgt = tf.py_function(encode,
+                                            [src, tgt],
+                                            [tf.int64, tf.int64])
+    result_src.set_shape([None])
+    result_tgt.set_shape([None])
+    return {'inputs': result_src, 'targets': result_tgt}
+
+  train_data = train_data.map(tf_encode, num_parallel_calls=AUTOTUNE)
+  eval_data = eval_data.map(tf_encode, num_parallel_calls=AUTOTUNE)
+
+  # Set bucketing.
+  if dynamic_batching:
+    train_buckets = None
+    eval_buckets = None
+  else:
+    # buckets should be (bucket_boundaries, bucket_batch_sizes)
+    train_buckets = ([max_target_length + 1], [batch_size])
```

Comment:
Why +1? I think this is explained in `bin_and_batch` as well, but I didn't really understand it. It could be useful to explain it here as well.

Reply:
again, tf.data's bucket_by_sequence_length pads to bucket_boundary -1, so we add the 1 to actually set it to desired length.